final updates to project 9

psswincher opened the pull request:
I completed git reset --hard to one of my earlier commits to make this pull request more complete, but have not been able to get main to go back far enough to capture all of my edits. That being said, the project-9 branch is complete and functioning.

psswincher committed:
refactor: class implementation

full project overhauled to take classes instead of the old spots function based code. Whew.

psswincher committed:
feat: api and edit avatar

implemented class API
user can now edit their avatar
api storage for cards, card like data
api storage for editing profile information (title and description)

psswincher committed:
feat: class api integration

full api implementation

psswincher committed:
final push

psswincher committed:
Revert "final push"

This reverts commit 0c30814bdc44a4a1f97a5a0e0ec9e8b531866add.

psswincher committed:
Revert "feat: class api integration"

This reverts commit 6f415e537d86bbabe60896dc6d8b50ef8fef2ac4.

psswincher committed:
Revert "feat: api and edit avatar"

This reverts commit 2f1185d8310229b789f0adc43a34b799e268b9f0.

psswincher committed:
Revert "refactor: class implementation"

This reverts commit 1f30a3bfbc03fba731f42cafe5def26bcab999ce.

psswincher committed:
final updates to project 9

before submission?

psswincher committed:
final updates to project 9

before submission?

gennady-bars reviewed (commented):
(**Please, expand the general comment ↓**)

Dear Preston, 

unfortunately I have to reject the project without reviewing:



* In **Files changed** there must be the code that should be reviewed. So, all changes from the last sprint should be there. But now you have only 1 file there. I'm not able to leave comments in your code. Please, try to send me a new pull request with all files changed. One of the options is to make a new **branch** from `main` then delete all code from `main`, commit it, push it and after that you need to make a pull request (so, there will be the whole code of the project)

Or you can try to remove the committed code from `main` like this:

1. `git checkout main`

2. find the commit that was the last one made while working on project 8

3. reset `main` branch to this commit with `git reset --hard commit-hash`

psswincher committed:
Merge branch 'project-9' of https://github.com/psswincher/se_project_spots into project-9

psswincher committed:
Prettier implementation and debugged adding new cards.

gennady-bars reviewed (commented):
(**Please, expand the general comment ↓**)

Dear Preston, 

unfortunately I have to reject the project without reviewing:



* When I like **new** cards, there are errors https://snipboard.io/o5tdBN.jpg 
* I can't close the confirmation modal by clicking on the close icon or `overlay`
* When I delete 3 cards in a row, there are errors https://snipboard.io/98ZDgJ.jpg 
* When I reload the page, all the new cards which I deleted a second ago, appear again on the page


I've noticed a mistake:

* New cards should be added only with `prepend` to the start of the section, otherwise the user doesn’t see the new card (if there are **6 or more** cards on the page) and may think that it hasn’t been created
* You need to **`disable`** the submit button in the `card form` if the inputs are **empty**. It’s needed because I can click the button and add an empty card, but it should **not** happen. After the first submitting the button gets activated, and you need to **deactivate** it again. Even if I can't add an empty card, the button should look **disabled** (gray) anyway and have attribute `disabled`.

psswincher commented:
Hello and thank you! I'll look into these; apologies for the repeat issues. I finally had time with a tutor to go back through git and restore my project to its original condition, and we had thought these issues were tied to git shenanigans and not bugs.

psswincher committed:
Debug: card renderer

Card renderer inadvertently was looping card adds, unshifting the top of the array to render the same card over and over.

New cards weren't properly being assigned ids. Fixed!

gennady-bars commented on src/components/Api.js:
```diff
@@ -0,0 +1,115 @@
+  request(routeData, body) {
+    const requestInfo = this._buildRequest(routeData, body);
+    return fetch(requestInfo.url, requestInfo.info)
+      .then((res) => {
+        if (res.ok) {
+          return res.json();
+        }
+        return Promise.reject(`API request Error: ${res.status}`);
```
```js
.then(res => {
                if (res.ok) {
                    return res.json();
                }
                return Promise.reject(`Error ${res.status}`);
            })

```

The code of `ok` checking is repeated in **every** request of the project. To avoid it you need to make a special method `_checkResponse` that will check any response from the server if it’s ok or not.


```js
_checkResponse(res) {
        // here is the code of the checking
    }

```

And now you can replace redundant repeating with the only line:


```js
 .then(this._checkResponse)

```

Please, pay attention that you need to pass only the **reference** for the method rather than the call. 

**Could be improved**

You can make a special method (function ) for fetching and checking responses **not to duplicate** it in every request:

```js

  _request(url, options) {
    return fetch(url, options).then(this._checkResponse)
  }

```

And now you can safely replace `fetch` with `this._request` and remove repeated checking. And you don’t need to change anything else.

vigilantebaratx replied:
Neat. Thank you!

gennady-bars commented on src/components/Api.js:
```diff
@@ -0,0 +1,115 @@
+        }
+        return Promise.reject(`API request Error: ${res.status}`);
+      })
+      .catch((err) => {
+        console.error(err); // log the error to the console
```
It’s better not to use `catch` in `Api` because you need to `catch` possible errors at the very **end** of any request, and this end is in `index.js` after all blocks `then`. And in real projects blocks `catch` usually open error popups and so on.

psswincher replied:
Got it! Thank you; I was unsure where to include catches and where to not include them, and so I clearly was including them at random. Glad to have this rule of them.

gennady-bars commented on src/components/Api.js:
```diff
@@ -0,0 +1,115 @@
+  requestDelete(cardId) {
+    return fetch(this._baseUrl + routes.deleteCardById.route + cardId, {
+      method: "DELETE",
+      headers: {
+        authorization: "0ba3486c-b117-4196-b741-be8eda3e197d",
+        "Content-Type": "application/json",
```
According to the task you need to pass full `headers` into the constructor of `Api`. See here https://snipboard.io/pRdaAZ.jpg . And in the code you need to use only: `headers: this._headers` (not to duplicate the code)

gennady-bars commented on src/components/CardElement.js:
```diff
@@ -0,0 +1,133 @@
+    if (this.isLikeActive()) {
+      this._onLikeAPIManager(this.id, false).then((result) => {
+        if (result) this.deactivateLikeButton();
+      });
```
You need to `catch` possible errors at the very **end** of any server request.

**Please, correct it everywhere**

gennady-bars commented on src/components/ModalWithForm.js:
```diff
@@ -0,0 +1,78 @@
+    );
+    this._inputFields = this._modalElement.querySelectorAll(".modal__input");
+    this._form = this._modalElement.querySelector(".modal__form");
+    this._currentProfileName = document.querySelector(".profile__name");
```
```js

   this._currentProfileName = document.querySelector(".profile__name");
    this._currentProfileDescription = document.querySelector(
      ".profile__description"
    );
```

Please, delete them because you don't use them

And `ModalWithForm` is a universal class for any modals with a form - it should not contain hardcoded data

gennady-bars commented on src/components/ModalWithForm.js:
```diff
@@ -0,0 +1,78 @@
+  }
+
+  _resetInputs() {
+    this._inputFields.forEach((field) => {
+      field.value = "";
```
You can easily clear the inputs with `reset`. Every **form** has this method

```js

  _resetInputs() {
    this._form.reset()
  }
```

psswincher replied:
Of course this already exists. I should have remembered. Thank you!

gennady-bars commented on src/components/ModalWithForm.js:
```diff
@@ -0,0 +1,78 @@
+    this._onSubmitCallback(this._getInputValues())
+      .then(() => {
+        this._resetInputs();
+        this._setSubmitToSubmit();
```
No need to repeat `this._setSubmitToSubmit();`  in `then` and `catch`. It should be just in `finally`

gennady-bars commented on src/components/ModalWithForm.js:
```diff
@@ -0,0 +1,78 @@
+        this.close();
+      })
+      .catch(() => {
+        this._setSubmitToSubmit();
```
No need to repeat `this._setSubmitToSubmit();`  in `then` and `catch`. It should be just in `finally`

gennady-bars commented on src/pages/index.js:
```diff
@@ -1,275 +1,242 @@
+          profileDescription: data.about,
+          imageUrl: data.avatar,
+        });
+        editProfileForm.setDefaultInputs(userProfile.getUserProfile());
```
`        editProfileForm.setDefaultInputs(userProfile.getUserProfile());
`

It should be on every opening 

* See here https://snipboard.io/OhCqu2.jpg You need to **insert** the profile data into the inputs when you open the **profile popup** to let the user **edit** the data and submit it. But after the 1st submition there is no data on opening

psswincher replied:
Fixed!

gennady-bars commented on src/components/ModalWithForm.js:
```diff
@@ -0,0 +1,78 @@
+      });
+  }
+
+  _setSubmitToSubmitting() {
+    this._submitButton.textContent = "Submitting...";
+  }
+
+  _setSubmitToSubmit() {
+    this._submitButton.textContent = "Submit";
```
**Could be improved**

It’s better to make method `renderLoading` inside class `PopupWithForm`. You should find the button once in the constructor and change the text:  

```js
  constructor({ popupSelector, handleSubmit }) {
    super(popupSelector);
    // fix the initial button text only once in the constructor
    this._submitBtnText = this._submitBtn.textContent
  }
  // add 2 params: isLoading and loadingText with a default text
  renderLoading(isLoading, loadingText='Saving...') {
    if (isLoading) {
      this._submitBtn.textContent = loadingText;
    } else {
	// here we return back the initial text. So, you don’t need to bother yourself about it
      this._submitBtn.textContent = this._submitBtnText;
    }
  }

```

gennady-bars commented on src/pages/index.js:
```diff
@@ -1,275 +1,242 @@
+const editProfileForm = new ModalWithForm({
+  modalSelector: modalSelectors.editProfileModal,
+  onSubmitCallback: (inputValues) => {
+    return new Promise((resolve, reject) => {
```
**Could be improved**

If it’s interesting for you here is how we can make a universal function for handling any submit. We can get rid of such duplicating as **loading effect**, **closing** and **catching errors**


```js

// You can make a universal function that accepts a request, popup instance and optional loading text  
function handleSubmit(request, popupInstance, loadingText = "Saving...") {

  // here we change the button text
  popupInstance.renderLoading(true, loadingText);
  request()
    .then(() => {
      // We need to close only in `then`
      popupInstance.close()
    })
      // we need to catch possible errors
      // console.error is used to handle errors if you don’t have any other ways for that
     .catch(console.error)
    // in `finally` we need to return the initial button text back in any case
    .finally(() => {
      popupInstance.renderLoading(false);
    });
}



```


You need to pass the **request** into the call of `handleSubmit` like that:



```js

// here is an example of the profile form handling
function handleProfileFormSubmit(inputValues) {
  // we create a function that returns a promise
  function makeRequest() {
    // `return` lets us use a promise chain `then, catch, finally` inside `handleSubmit`
    return api.editProfile(inputValues).then((userData) => {
      userInfo.setUserInfo(userData)
    });
  }
  // Here we call the function passing the request, popup instance and if we need some other loading text we can pass it as the 3rd argument
  handleSubmit(makeRequest, profilePopup);
}



```

psswincher replied:
That is super cool. Thank you! Excited to try and implement this.

gennady-bars commented on src/pages/index.js:
```diff
@@ -1,275 +1,242 @@
-    // checkFormFieldInputErrors(modal.querySelector('.modal__form')); //this can check every field on open, but feels bearish
-    modal.classList.add('modal_open');
-}
+const newPostFormValidator = new FormValidator(
```
**Could be improved**

You can **universally** create instances of validators for all forms in the project storing them inside one object: `formValidators`.  Then you can take any validator using attribute `name` of the form where you need to disable/enable the submit button or remove errors.

```js
// define an object for storing validators
const formValidators = {}

const enableValidation = (config) => {
  const formList = Array.from(document.querySelectorAll(config.formSelector))
  formList.forEach((formElement) => {
    const validator = new FormValidator(config, formElement)
    // Here you get the name of the form (if you don’t have it then you need to add it into each form in `index.html` first)
    const formName = formElement.getAttribute('name')

   // Here you store the validator using the `name` of the form
    formValidators[formName] = validator;
    validator.enableValidation();
  });
};

enableValidation(config);

```

And now you can use them for disabling buttons or clearing errors:

```js

formValidators[ profileForm.getAttribute('name') ].resetValidation()

// or you can use a string – the name of the form (you know it from `index.html`)

formValidators['profile-form'].resetValidation()

```

psswincher replied:
Sooooo much tidier. Thank you!

gennady-bars reviewed (commented):
**↓Expand to see the whole summary↓**

Hello, Preston!

Great job on submitting your project!  



There are few things that need to be corrected in your project. They're mostly minor issues that are easy to fix.

**Needs correcting:**

* All comments in the code and the comments below:
* You need to **`disable`** the submit button in the `card form` if the inputs are **empty**. It’s needed because I can click the button and add an empty card, but it should **not** happen. After the first submitting the button gets activated, and you need to **deactivate** it again. Even if I can't add an empty card, the button should look **disabled** (gray) anyway and have attribute `disabled`.
* See here https://snipboard.io/OhCqu2.jpg You need to **insert** the profile data into the inputs when you open the **profile popup** to let the user **edit** the data and submit it. But after the 1st submition there is no data on opening
* The code of `ok` checking is repeated in **every** request of the project. To avoid it you need to make a special method `_checkResponse` that will check any response from the server if it’s ok or not.
* It’s better not to use `catch` in `Api` because you need to `catch` possible errors at the very **end** of any request, and this end is in `index.js` after all blocks `then`. And in real projects blocks `catch` usually open error popups and so on.
* According to the task you need to pass full `headers` into the constructor of `Api`. See here https://snipboard.io/pRdaAZ.jpg . And in the code you need to use only: `headers: this._headers` (not to duplicate the code)
* You need to `catch` possible errors at the very **end** of any server request.


There are some comments in your project that can help you make it even better. Please check the `Could be improved` comments.

**A reminder:** It’s necessary to fix all  comments to get the project accepted. And `Could be improved` comments are optional, they will be up to you if you want to implement them.

Your hard work on this project is truly commendable. While there are still some adjustments needed. If you need any assistance with the feedback or understanding certain concepts, our tutors are here to help. Keep moving forward and maintain your dedication — you are doing an amazing job!✨

psswincher committed:
Refactor: handleSubmit, error handling

Debugged form handling to ensure blank posts can't be created.

Thank you Gennady! A good time was had by all.

gennady-bars commented on src/pages/index.js:
```diff
@@ -1,275 +1,233 @@
+          avatar: inputValues[userProfileSettings.profileInputAvatarLink],
+        })
+        .then(() => {
+          userProfile.setUserImage(
```
* The avatar submit button should be disabled also after the form is reset in `then` https://snipboard.io/7jYl2K.jpg 


`resetValidation` should be called here

psswincher replied:
I moved resetValidation into the handleSubmit function; all forms will now have their validators reset automatically on submit, no need to repeat for each form? Submitting!

gennady-bars reviewed (commented):
Everything is almost fine, but:


* The avatar submit button should be disabled also after the form is reset in `then` https://snipboard.io/7jYl2K.jpg

psswincher committed:
Merge branch 'project-9' of https://github.com/psswincher/se_project_spots into project-9

psswincher committed:
Fix: submit form validation bug

I moved resetValidation into the handleSubmit function; all forms will now have their validators reset automatically on submit, no need to repeat each time.

gennady-bars commented on src/pages/index.js:
```diff
@@ -229,5 +229,7 @@ function handleSubmit(request, modalInstance) {
     .catch(console.error)
-    .finally(() => modalInstance.renderLoading(false));
+    .finally(() => {
+        formValidators[modalInstance.formName].resetValidation();
```
`  formValidators[modalInstance.formName].resetValidation();` should be in `then` to disable the button only after a successful response to be able to click the button in case of a server error

gennady-bars commented on src/pages/index.js:
```diff
@@ -229,5 +229,7 @@ function handleSubmit(request, modalInstance) {
     .catch(console.error)
-    .finally(() => modalInstance.renderLoading(false));
+    .finally(() => {
+        formValidators[modalInstance.formName].resetValidation();
```
* `deleteConfirmationManager` doesn't have a validator, so you need to protect your code from errors.


It's done like this `formValidators[modalInstance.formName]?.resetValidation();` . You should use `?` before the **dot**

```js
function handleSubmit(request, modalInstance) {
  modalInstance.renderLoading(true);
  request()
    .then(() => {
      modalInstance.close();
      formValidators[modalInstance.formName]?.resetValidation();  // here
    })
    .catch(console.error)
    .finally(() => {
      modalInstance.renderLoading(false);
    });
}


```

psswincher replied:
I accomplished this by refactoring the ModalConfirm class. I moved the logic for handling the delete out of the makeNewCard function and into the deleteConfirmationManager declaration. 
By doing so, I could:
- rope ModalConfirm's handling into the handleSubmit function, drying these up quite a bit
- clean up the ModalConfirm class to be way more readable/simple

gennady-bars commented on src/pages/index.js:
```diff
@@ -122,7 +123,6 @@ const newPostForm = new ModalWithForm({
-          formValidators[newPostForm.formName].resetValidation();
           cardSection.createAndRenderItem(res);
         });
     }
```
And `onDeleteAPIManager` should be changed a bit. You should `return` the promise from the **api** call like this


```js


  const onDeleteAPIManager = (deleteConfirmationManager, id) => {
    function makeRequest() {
      return deleteConfirmationManager.awaitUserChoice().then((choice) => {
        if (choice) {
// here you return the promise to be able to see `Deletin...` in the button
          return dbApi.requestDelete(id).then(() => {
            newCard.remove();
          });
        }
      });
    }

```

gennady-bars reviewed (commented):
One more step is needed:

* `  formValidators[modalInstance.formName].resetValidation();` should be in `then` to disable the button only after a successful response to be able to click the button in case of a server error
* `deleteConfirmationManager` doesn't have a validator, so you need to protect your code from errors.

psswincher committed:
refactor: onConfirm validation

gennady-bars reviewed (approved):
You’ve done a great job!

Your project has been accepted. 

Good luck!
